proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +144,500 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
+
+  details::__builtin_MatVecMul(
+      /*out*/ OutputVector, details::IsUnsigned<OutputElTy>(), InputVector.Data,
+      details::IsUnsigned<InputElTy>(), InputDT, BUFFER_HANDLE(Matrix.Buffer),
+      Matrix.StartOffset, MatrixDT, MatrixM, MatrixK, MatrixLayout,
+      MatrixTranspose, Matrix.Stride);
+
+  return OutputVector;
+}
+
+} // namespace linalg
+} // namespace dx
+```
+
+## Function: MulAdd
+
+The `dx::linalg::MulAdd` function behaves as `dx::linalg::Mul`, but also adds a
+bias vector (loaded from memory) to the result.
+
+Example:
+
+```c++
+ByteAddressBuffer Buffer;
+
+void Example() {
+  using namespace dx::linalg;
+
+  MatrixRef<DATA_TYPE_E4M3, 32, 8, MATRIX_LAYOUT_MUL_OPTIMAL> Matrix = {Buffer,
+                                                                        0, 0};
+
+  VectorRef<DATA_TYPE_FLOAT16> BiasVector = {Buffer, 1024};
+
+  vector<float, 8> V = 0;
+  vector<float, 32> Result =
+      MulAdd<float>(Matrix, InterpretedVector<DATA_TYPE_E4M3>(V), BiasVector);
+}
+```
+
+Conceptual API:
+
+```c++
+template<typename TYo>
+vector<TYo, M_M> Mul(MatrixRefImpl<...> Matrix, Vector<...> InputVector, VectorRefImpl<...> BiasVector);
+```
+
+See [0029] for details of this operation.
+
+> TODO: details of what this function does really belong in here as well, but
+> for now the source-of-truth is [0029].
+
+Implementation:
+
+```c++
+template <typename OutputElTy, typename InputElTy, int InputElCount,
+          typename MatrixBufferTy, DataType InputDT, DataType MatrixDT,
+          uint MatrixM, uint MatrixK, MatrixLayout MatrixLayout,
+          bool MatrixTranspose, typename BiasVectorBufferTy,
+          DataType BiasVectorDT>
+vector<OutputElTy, MatrixM>
+MulAdd(MatrixRefImpl<MatrixBufferTy, MatrixDT, MatrixM, MatrixK, MatrixLayout,
+                     MatrixTranspose>
+           Matrix,
+       Vector<InputElTy, InputElCount, InputDT> InputVector,
+       VectorRefImpl<BiasVectorBufferTy, BiasVectorDT> BiasVector) {
+
+  vector<OutputElTy, MatrixM> OutputVector;
+
+  details::__builtin_MatVecMulAdd(
+      /*out*/ OutputVector, details::IsUnsigned<OutputElTy>(), InputVector.Data,
+      details::IsUnsigned<InputElTy>(), InputDT, BUFFER_HANDLE(Matrix.Buffer),
+      Matrix.StartOffset, MatrixDT, MatrixM, MatrixK, MatrixLayout,
+      MatrixTranspose, Matrix.Stride, BUFFER_HANDLE(BiasVector.Buffer),
+      BiasVector.StartOffset, BiasVectorDT);
+
+  return OutputVector;
+}
+```
+
+## Function: OuterProductAccumulate
+
+Computes the outer product between column vectors and an **M**x**N** matrix is
+accumulated component-wise atomically (with device scope) in memory.
+
+The operation is equivalent to:
+
+> ResultMatrix = InputVector1 \* Transpose(InputVector2);
+
+Example:
+
+```c++
+RWByteAddressBuffer RWBuf;
+
+void Example(vector<half, 128> Input1, vector<half, 256> Input2) {
+  using namespace dx::linalg;
+
+  RWMatrixRef<DATA_TYPE_FLOAT16, 128, 256, MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL>
+      matrix = {RWBuf, 0, 0};
```

Comment:
Should this be 'M'atrix?

Reply:
Yes - I've done a pass through all the variable naming to try and make sure it's consistent.  Please continue to call out inconsistencies!